synapse/federation/federation_client.py

```diff
@@ -358,54 +359,97 @@ async def get_pdu(
             The requested PDU, or None if we were unable to find it.
         """
```

Comment:
> I have a suggestion for a somewhat different approach. It's arguably cleaner, but also probably a bunch more work; I'd be interested in your thoughts.
> 
> In my ideal world, `_EventInternalMetadata.outlier` should be a *private*, *immutable* property. Which is to say, you have to know whether you are dealing with an outlier at the time you construct an `_EventInternalMetadata` and hence when you call `make_event_from_dict`. We're some way off that goal in the codebase as a whole, but that doesn't stop us thinking about how to move in that direction for this part of the code.
> 
> *-- https://github.com/matrix-org/synapse/pull/13320#issuecomment-1188873153*

I like the idea of having it be immutable 👍

> It would mean passing an `outlier` flag into `get_pdu` telling it whether we're going to persist the result as an outlier or not. (That's fairly easy, because `get_pdu` is called in two places: `FederationEventHandler.get_event`, which only deals in outliers, and `FederationClient._check_sigs_and_hash_and_fetch_one`, where we can infer the outlieryness from the input `pdu`).
>
> That then raises a few possibilities: we could continue to construct a new `EventBase` on each call to `get_pdu` as proposed here. Or we could use a different cache key depending on the outlier flag and have (up to) two cache entries for each event.

But needing to pass in `outlier` or the metadata (when we need to add more than just `outlier` in the future) makes the function signature surface a bit icky. I'd rather just force downstream consumers to create a copy with the metadata they need/expect, or want to add.

We can tackle this in another potential PR though.

Reply:
yes, there's definitely a balance of two evils here. I'm happy to punt this for now at least.